data_model/src/predicate.rs

```diff
@@ -0,0 +1,872 @@
+    fn applies(&self, input: &Value) -> bool {
+        match self {
+            PredicateBox::And(vector) => vector.iter().all(|pred| pred.applies(input)),
+            PredicateBox::Or(vector) => vector.iter().any(|pred| pred.applies(input)),
+            PredicateBox::Not(predicate) => !predicate.applies(input),
+            PredicateBox::Raw(predicate) => predicate.applies(input),
+        }
+    }
+}
+
+#[cfg(test)]
+pub mod test {
+    #![allow(clippy::print_stdout, clippy::use_debug)]
+
+    use super::{value, PredicateBox};
+    use crate::{PredicateTrait as _, Value};
+
+    #[test]
+    fn pass() {
+        let t = PredicateBox::new(value::Predicate::Pass);
+        let f = t.clone().negate();
+        let v_t = Value::from(true);
+        let v_f = Value::from(false);
+        println!("t: {t:?}, f: {f:?}");
+
+        assert!(t.applies(&v_t));
+        assert!(t.applies(&v_f));
+        assert!(!f.applies(&v_t));
+        assert!(!f.applies(&v_f));
+    }
+
+    #[test]
+    fn truth_table() {
+        let t = PredicateBox::new(value::Predicate::Pass);
+        let f = t.clone().negate();
+        let v = Value::from(true);
+
+        assert!(!PredicateBox::and(t.clone(), f.clone()).applies(&v));
+        assert!(PredicateBox::and(t.clone(), t.clone()).applies(&v));
+        assert!(!PredicateBox::and(f.clone(), f.clone()).applies(&v));
+        assert!(!PredicateBox::and(f.clone(), t.clone()).applies(&v));
+
+        assert!(PredicateBox::or(t.clone(), t.clone()).applies(&v));
+        assert!(PredicateBox::or(t.clone(), f.clone()).applies(&v));
+        assert!(PredicateBox::or(f.clone(), t).applies(&v));
+        assert!(!PredicateBox::or(f.clone(), f).applies(&v));
+    }
+
+    #[test]
+    fn negation() {
+        let t = PredicateBox::default();
+
+        assert!(matches!(t.clone().negate().negate(), PredicateBox::Raw(_)));
+        // De-morgan identities
+        assert!(matches!(
+            PredicateBox::and(t.clone(), t.clone()).negate(),
+            PredicateBox::Or(_)
+        ));
+        assert!(matches!(
+            PredicateBox::or(t.clone(), t).negate(),
+            PredicateBox::And(_)
+        ));
+    }
+}
+
+pub mod string {
+    //! String-related predicates and implementations.
+    use super::*;
+
+    /// Predicate useful for processing [`String`]s and [`Name`]s.
+    #[derive(Debug, Clone, Serialize, Deserialize, Encode, Decode, IntoSchema)]
+    pub enum Predicate {
+        /// Forward to [`str::contains()`]
+        Contains(String),
+        /// Forward to [`str::starts_with()`]
+        StartsWith(String),
+        /// Forward to [`str::ends_with()`]
+        EndsWith(String),
+        /// Forward to [`String`] equality.
+        Is(String),
+    }
+
+    impl Predicate {
+        /// Construct the [`Self::Contains`] variant
+        #[inline]
+        pub fn contains(predicate: &str) -> Self {
```

Comment:
Why not something like

```suggestion
        pub fn contains(predicate: impl Into<String>) -> Self {
```

Reply:
Yeah, I get that. What I meant is why not also accept owned `String` in short-hand functions?

Reply:
I might not understand what is being proposed here but I think `impl Into<String>` is something you'll almost never need. It's just complicating the API. Function argument should either be owned `String` or `&str` slice. Converting structures into String is usually done just for display

What maybe could be done is to make use of the `Borrow` trait but that seems too restrictive

Reply:
> Function argument should either be owned `String` or `&str` slice.

I'm proposing accepting both here. 
I don't actually have a strong opinion on this anyway, so feel free to mark resolved.

Reply:
Again, if you want to use a string slice, you use the lowercase function and `.to_owned()` is implied. 

If you want to use the owned string object, you construct/`into()` in place using the upper-case function. 

We support both uses.